<a href="https://colab.research.google.com/github/Meenu2805/PortfolioProjects/blob/main/Data_Cleaning_with_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Automobile data set consists of three types of entities:
(a) the specification of an auto in terms of various characteristics,
(b) its assigned insurance risk rating,
(c) its normalized losses in use as compared to other cars. The second rating corresponds to the degree to which the auto is more risky than its price indicates. Cars are initially assigned a risk factor symbol associated with its price. Then, if it is more risky (or less), this symbol is adjusted by moving it up (or down) the scale. Actuarians call this process "symboling". A value of +3 indicates that the auto is risky, -3 that it is probably pretty safe. The third factor is the relative average loss payment per insured vehicle year. This value is normalized for all autos within a particular size classification (two-door small, station wagons, sports/specialty, etc…), and represents the average loss per car per year.

Number of Instances: 205

Number of Attributes: 26 total
-- 15 continuous
-- 1 integer
-- 10 nominal.

Overview:

>>Importing the necessary libraries and the raw data.

>>Finding the missing values.

>>Filling the missing values.

>>Summary and Conclusion.


In [ ]:
#Importing the necessary libraries and the raw data
import pandas as pd
import numpy as np

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving Automobile price data _Raw_.csv to Automobile price data _Raw_.csv


In [ ]:
import io
df = pd.read_csv(io.BytesIO(uploaded['Automobile price data _Raw_.csv']))
# Dataset is now stored in a Pandas Dataframe

In [ ]:
df.head

<bound method NDFrame.head of      symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            3                 ?  alfa-romero       gas        std   
2            1                 ?  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd        

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

1) If we closely look at the data we will find that there are several '?' sign on the data points which indicates that there are some missing values in the dataset.
2) As it is difficult to deal with '?' in tha dataset we will replacing it with 'NaN' value so that we can easily deal with the null values.

In [ ]:
#Finding the missing values
df[df == '?'] #to find the location of all '?' in the dataset

symboling normalized-losses make fuel-type aspiration num-of-doors  \
0          NaN                 ?  NaN       NaN        NaN          NaN   
1          NaN                 ?  NaN       NaN        NaN          NaN   
2          NaN                 ?  NaN       NaN        NaN          NaN   
3          NaN               NaN  NaN       NaN        NaN          NaN   
4          NaN               NaN  NaN       NaN        NaN          NaN   
..         ...               ...  ...       ...        ...          ...   
200        NaN               NaN  NaN       NaN        NaN          NaN   
201        NaN               NaN  NaN       NaN        NaN          NaN   
202        NaN               NaN  NaN       NaN        NaN          NaN   
203        NaN               NaN  NaN       NaN        NaN          NaN   
204        NaN               NaN  NaN       NaN        NaN          NaN   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0          NaN          NaN             NaN         NaN  ...          NaN   
1          NaN          NaN             NaN         NaN  ...          NaN   
2          NaN          NaN             NaN         NaN  ...          NaN   
3          NaN          NaN             NaN         NaN  ...          NaN   
4          NaN          NaN             NaN         NaN  ...          NaN   
..         ...          ...             ...         ...  ...          ...   
200        NaN          NaN             NaN         NaN  ...          NaN   
201        NaN          NaN             NaN         NaN  ...          NaN   
202        NaN          NaN             NaN         NaN  ...          NaN   
203        NaN          NaN             NaN         NaN  ...          NaN   
204        NaN          NaN             NaN         NaN  ...          NaN   

     fuel-system  bore  stroke compression-ratio horsepower  peak-rpm  \
0            NaN   NaN     NaN               NaN        NaN       NaN   
1            NaN   NaN     NaN               NaN        NaN       NaN   
2            NaN   NaN     NaN               NaN        NaN       NaN   
3            NaN   NaN     NaN               NaN        NaN       NaN   
4            NaN   NaN     NaN               NaN        NaN       NaN   
..           ...   ...     ...               ...        ...       ...   
200          NaN   NaN     NaN               NaN        NaN       NaN   
201          NaN   NaN     NaN               NaN        NaN       NaN   
202          NaN   NaN     NaN               NaN        NaN       NaN   
203          NaN   NaN     NaN               NaN        NaN       NaN   
204          NaN   NaN     NaN               NaN        NaN       NaN   

    city-mpg highway-mpg price  
0        NaN         NaN   NaN  
1        NaN         NaN   NaN  
2        NaN         NaN   NaN  
3        NaN         NaN   NaN  
4        NaN         NaN   NaN  
..       ...         ...   ...  
200      NaN         NaN   NaN  
201      NaN         NaN   NaN  
202      NaN         NaN   NaN  
203      NaN         NaN   NaN  
204      NaN         NaN   NaN  

[205 rows x 26 columns]

In [ ]:
#Now we can replace '?' with Nan values
df = df.replace('?', np.NaN)
df

symboling normalized-losses         make fuel-type aspiration  \
0            3               NaN  alfa-romero       gas        std   
1            3               NaN  alfa-romero       gas        std   
2            1               NaN  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

In [ ]:
#Since now all the '?' is successfully replaced with Nan, we can find out the number of missing values

In [ ]:
df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

By observing the above results we have successfully counted the number missing values in different columns.

According to the results,

a) Normalized Losses has 40 missing values.
b) No. of Doors has 2 missing values.
c) Bore has 4 missing values.
d) Stroke has 4 missing values.
e) Horsepower and Peak RPM have 2 missing values each.
f) Price has 4 missing values.

We are now going to check for the number of rows having at least one missing values.
For this process, we will first find the indexes of the datasets which contains at least one null values and then pass it on to the dataframe.

In [ ]:
d = df.isnull().any(axis = 1) #axis = 1 for rows
index = d.index[d.values]
index

Int64Index([  0,   1,   2,   5,   7,   9,  14,  15,  16,  17,  27,  43,  44,
             45,  46,  48,  49,  55,  56,  57,  58,  63,  66,  71,  73,  74,
             75,  82,  83,  84, 109, 110, 113, 114, 124, 126, 127, 128, 129,
            130, 131, 181, 189, 191, 192, 193],
           dtype='int64')

These are the indexes of the dataset having at least one missing value throughout the datapoint.

Now we are going to locate all these datapoints at once and counting it.

In [ ]:
missing = df.iloc[index, :] #no of rows having atleast one missing value
missing

symboling normalized-losses           make fuel-type aspiration  \
0            3               NaN    alfa-romero       gas        std   
1            3               NaN    alfa-romero       gas        std   
2            1               NaN    alfa-romero       gas        std   
5            2               NaN           audi       gas        std   
7            1               NaN           audi       gas        std   
9            0               NaN           audi       gas      turbo   
14           1               NaN            bmw       gas        std   
15           0               NaN            bmw       gas        std   
16           0               NaN            bmw       gas        std   
17           0               NaN            bmw       gas        std   
27           1               148          dodge       gas      turbo   
43           0               NaN          isuzu       gas        std   
44           1               NaN          isuzu       gas        std   
45           0               NaN          isuzu       gas        std   
46           2               NaN          isuzu       gas        std   
48           0               NaN         jaguar       gas        std   
49           0               NaN         jaguar       gas        std   
55           3               150          mazda       gas        std   
56           3               150          mazda       gas        std   
57           3               150          mazda       gas        std   
58           3               150          mazda       gas        std   
63           0               NaN          mazda    diesel        std   
66           0               NaN          mazda    diesel        std   
71          -1               NaN  mercedes-benz       gas        std   
73           0               NaN  mercedes-benz       gas        std   
74           1               NaN  mercedes-benz       gas        std   
75           1               NaN        mercury       gas      turbo   
82           3               NaN     mitsubishi       gas      turbo   
83           3               NaN     mitsubishi       gas      turbo   
84           3               NaN     mitsubishi       gas      turbo   
109          0               NaN         peugot       gas        std   
110          0               NaN         peugot    diesel      turbo   
113          0               NaN         peugot       gas        std   
114          0               NaN         peugot    diesel      turbo   
124          3               NaN       plymouth       gas      turbo   
126          3               NaN        porsche       gas        std   
127          3               NaN        porsche       gas        std   
128          3               NaN        porsche       gas        std   
129          1               NaN        porsche       gas        std   
130          0               NaN        renault       gas        std   
131          2               NaN        renault       gas        std   
181         -1               NaN         toyota       gas        std   
189          3               NaN     volkswagen       gas        std   
191          0               NaN     volkswagen       gas        std   
192          0               NaN     volkswagen    diesel      turbo   
193          0               NaN     volkswagen       gas        std   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
5            two        sedan          fwd           front        99.8  ...   
7           four        wagon          fwd           front       105.8  ...   
9            two    hatchback          4wd           front        99.5  ...   
14          four        sedan          rwd           front       10

By this we can see that there are 46 rows having atleast 1 missing value

Now we are going to check if there is any row or column having all null values.
So that we can delete that entire row or column

In [ ]:
df.isnull().all(axis=0).sum() #for columns having all missing value

0

In [ ]:
df.isnull().all(axis=1).sum() #for rows having all missing values

0

So from the above command we got to know that there is no column or row exist which has all the null values.

Now, let's find out the percentage of missing value each column holds

In [ ]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100,2)
#we got all the columns missing value percentage on the top using 'ascending = False'

normalized-losses    20.00
price                 1.95
stroke                1.95
bore                  1.95
peak-rpm              0.98
num-of-doors          0.98
horsepower            0.98
engine-type           0.00
highway-mpg           0.00
city-mpg              0.00
compression-ratio     0.00
fuel-system           0.00
engine-size           0.00
num-of-cylinders      0.00
symboling             0.00
height                0.00
width                 0.00
length                0.00
wheel-base            0.00
engine-location       0.00
drive-wheels          0.00
body-style            0.00
aspiration            0.00
fuel-type             0.00
make                  0.00
curb-weight           0.00
dtype: float64

According to the results it has been found that:

Normalized Losses is the column having the most null values of about 20.00 %.
Price,Stroke and Bore are the columns containing 1.95% null values each.
Peak RPM,Horsepower and No. of Doors are the columns containing 0.98 % null values each.

In [ ]:
#Filling the missing values

Now we know every detail of the missing values in the dataset and we will now fill the null values.

We will replace the null values with the mean of the values of that particular column.

In order to do that, the data type of the particular columns should be in floating types, otherwise we will not be able to find the mean of that particular column.

Changing the data types of the target columns.

**normalized-losses    **
**price  **               
**stroke  **              
**bore     **             
**peak-rpm   **     
num-of-doors          
**horsepower **

As '**No. of Doors**' column is a categorical data it can't be change into float data type. We will deal with this column later.

First we will change the data type of the target columns excluding 'No. of Doors'.

In [ ]:
df[['normalized-losses','price','stroke','bore','peak-rpm','horsepower']] = df[['normalized-losses','price','stroke','bore','peak-rpm','horsepower']].apply(pd.to_numeric)


In [ ]:
#now since we tried changing the dtype of these columns, so let's check if we can calculate their mean or not
df[['normalized-losses','price','stroke','bore','peak-rpm','horsepower']].describe()

normalized-losses         price      stroke        bore     peak-rpm  \
count         164.000000    201.000000  201.000000  201.000000   203.000000   
mean          122.000000  13207.129353    3.255423    3.329751  5125.369458   
std            35.442168   7947.066342    0.316717    0.273539   479.334560   
min            65.000000   5118.000000    2.070000    2.540000  4150.000000   
25%            94.000000   7775.000000    3.110000    3.150000  4800.000000   
50%           115.000000  10295.000000    3.290000    3.310000  5200.000000   
75%           150.000000  16500.000000    3.410000    3.590000  5500.000000   
max           256.000000  45400.000000    4.170000    3.940000  6600.000000   

       horsepower  
count  203.000000  
mean   104.256158  
std     39.714369  
min     48.000000  
25%     70.000000  
50%     95.000000  
75%    116.000000  
max    288.000000

So it is clear that we can find out the mean of the each target columns.

Replacing the null values in 'Normalized Losses' column with the mean of the column i.e. 122.

Let us deal with the 'Normalized Losses' column first as it contains the most number of null values. We will replace the null values with its mean which is 122.

In [ ]:
df.loc[:, 'normalized-losses'].fillna(df['normalized-losses'].mean(),inplace = False) #inplace=False (default) used to get the new copy of data

0      122.0
1      122.0
2      122.0
3      164.0
4      164.0
       ...  
200     95.0
201     95.0
202     95.0
203     95.0
204     95.0
Name: normalized-losses, Length: 205, dtype: float64

It looks like all the null values are replaced with the mean value of the column.

Let's check whether actually all the null values are gone or not.

In [ ]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100,2)

price                1.95
stroke               1.95
bore                 1.95
peak-rpm             0.98
num-of-doors         0.98
horsepower           0.98
engine-type          0.00
highway-mpg          0.00
city-mpg             0.00
compression-ratio    0.00
fuel-system          0.00
engine-size          0.00
num-of-cylinders     0.00
symboling            0.00
normalized-losses    0.00
height               0.00
width                0.00
length               0.00
wheel-base           0.00
engine-location      0.00
drive-wheels         0.00
body-style           0.00
aspiration           0.00
fuel-type            0.00
make                 0.00
curb-weight          0.00
dtype: float64

It is clear from the abpve results that there is no null value left in the 'Normalized Losses' column.

Similarly performing the same procedure with the rest of the target columns.
We will apply the same method for filling the null values as we did in the previous step.

In [ ]:
df.loc[:, 'price'].fillna(df['price'].mean(),inplace = True)
df.loc[:, 'stroke'].fillna(df['stroke'].mean(),inplace = True)
df.loc[:, 'bore'].fillna(df['bore'].mean(),inplace = True)
df.loc[:, 'peak-rpm'].fillna(df['peak-rpm'].mean(),inplace = True)
df.loc[:, 'horsepower'].fillna(df['horsepower'].mean(),inplace = True)


Now, let's check again if all the null values are gone or not

In [ ]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100,2)

num-of-doors         0.98
symboling            0.00
engine-type          0.00
highway-mpg          0.00
city-mpg             0.00
peak-rpm             0.00
horsepower           0.00
compression-ratio    0.00
stroke               0.00
bore                 0.00
fuel-system          0.00
engine-size          0.00
num-of-cylinders     0.00
curb-weight          0.00
normalized-losses    0.00
height               0.00
width                0.00
length               0.00
wheel-base           0.00
engine-location      0.00
drive-wheels         0.00
body-style           0.00
aspiration           0.00
fuel-type            0.00
make                 0.00
price                0.00
dtype: float64

The null values have been filled successfully except in 'No. of Doors' column.

Filling the null values in 'No. of Doors' columns.

In this case the data the data is categorical so we first need to look at the data finding the unique values of it.

In [ ]:
df['num-of-doors'].unique()

array(['two', 'four', nan], dtype=object)

So it is clear from the above results that there are two unique values under which the column is classified, that is 'two' and 'four'

As it is a categorical data the 'mean' method won't work here. In this case we need to count the number of these categorical indicators. The indicator which has more number of counts will be used in filling the null value.

In [ ]:
df['num-of-doors'].astype('category').value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

The above result indicates that there are more number of cars having 'four' doors than 'two' doors. So, accordingly we will use the 'four' indicator to fill the null values.

In [ ]:
df.loc[:, 'num-of-doors'].fillna('four', inplace = True)

In [ ]:
df['num-of-doors'].astype('category').value_counts()

four    116
two      89
Name: num-of-doors, dtype: int64

#We have successfully removed all the null values and hence we have cleaned our whole dataset in this way.

6. Summary and Conclusion:

>>The raw data was imported.

>>The null values was counted.

>>All the null values was replaced by the columns's mean value.

>>In the 'num-of-doors' column the null values are replaced by the most numbered categorical indicator.


Hence the 'Automobile Dataset' is cleaned.
